diff --git a/apps/openmw/mwrender/postprocessor.cpp b/apps/openmw/mwrender/postprocessor.cpp
index b6f72d3a64..6678360f2e 100644
--- a/apps/openmw/mwrender/postprocessor.cpp
+++ b/apps/openmw/mwrender/postprocessor.cpp
@@ -408,11 +408,14 @@ namespace MWRender
 
             if (mNormalsSupported)
             {
+/*
                 auto& shaderManager
                     = MWBase::Environment::get().getResourceSystem()->getSceneManager()->getShaderManager();
                 auto defines = shaderManager.getGlobalDefines();
                 defines["disableNormals"] = mNormals ? "0" : "1";
                 shaderManager.setGlobalDefines(defines);
+*/
+                mRendering.setWriteNormals(mNormals);
             }
 
             mRendering.getLightRoot()->setCollectPPLights(mPassLights);
diff --git a/apps/openmw/mwrender/renderingmanager.cpp b/apps/openmw/mwrender/renderingmanager.cpp
index 67d1cfb587..80d79bcad6 100644
--- a/apps/openmw/mwrender/renderingmanager.cpp
+++ b/apps/openmw/mwrender/renderingmanager.cpp
@@ -213,6 +213,7 @@ namespace MWRender
             : mFogStart(0.f)
             , mFogEnd(0.f)
             , mWireframe(false)
+            , mWriteNormals(false)
         {
         }
 
@@ -231,6 +232,11 @@ namespace MWRender
             }
             else
                 stateset->removeAttribute(osg::StateAttribute::POLYGONMODE);
+
+            stateset->setDefine("WRITE_NORMALS", (mWriteNormals) ? "1" : "0", osg::StateAttribute::ON);
+            stateset->setDefine("FORCE_PPL", (Settings::shaders().mForcePerPixelLighting == true) ? "1" : "0", osg::StateAttribute::ON);
+            stateset->setDefine("CLASSIC_FALLOFF", (Settings::shaders().mClassicFalloff == true) ? "1" : "0", osg::StateAttribute::ON);
+            stateset->setDefine("MAX_LIGHTS", std::to_string(Settings::shaders().mMaxLights), osg::StateAttribute::ON);
         }
 
         void apply(osg::StateSet* stateset, osg::NodeVisitor*) override
@@ -242,6 +248,8 @@ namespace MWRender
             fog->setColor(mFogColor);
             fog->setStart(mFogStart);
             fog->setEnd(mFogEnd);
+
+            stateset->setDefine("WRITE_NORMALS", (mWriteNormals) ? "1" : "0", osg::StateAttribute::ON);
         }
 
         void setAmbientColor(const osg::Vec4f& col) { mAmbientColor = col; }
@@ -263,12 +271,15 @@ namespace MWRender
 
         bool getWireframe() const { return mWireframe; }
 
+        void setWriteNormals(float enabled) {mWriteNormals = enabled; }
+
     private:
         osg::Vec4f mAmbientColor;
         osg::Vec4f mFogColor;
         float mFogStart;
         float mFogEnd;
         bool mWireframe;
+        bool mWriteNormals;
     };
 
     class PreloadCommonAssetsWorkItem : public SceneUtil::WorkItem
@@ -1541,17 +1552,20 @@ namespace MWRender
             else if (it->first == "Shaders"
                 && (it->second == "force per pixel lighting" || it->second == "classic falloff"))
             {
+/*
                 mViewer->stopThreading();
 
                 auto defines = mResourceSystem->getSceneManager()->getShaderManager().getGlobalDefines();
                 defines["forcePPL"] = Settings::shaders().mForcePerPixelLighting ? "1" : "0";
                 defines["classicFalloff"] = Settings::shaders().mClassicFalloff ? "1" : "0";
                 mResourceSystem->getSceneManager()->getShaderManager().setGlobalDefines(defines);
-
+*/
                 if (MWMechanics::getPlayer().isInCell() && it->second == "classic falloff")
                     configureAmbient(*MWMechanics::getPlayer().getCell()->getCell());
 
-                mViewer->startThreading();
+//                mViewer->startThreading();
+
+                mStateUpdater->reset();
             }
             else if (it->first == "Shaders"
                 && (it->second == "light bounds multiplier" || it->second == "maximum light distance"
@@ -1564,18 +1578,18 @@ namespace MWRender
 
                 if (it->second == "max lights" && !lightManager->usingFFP())
                 {
-                    mViewer->stopThreading();
+//                    mViewer->stopThreading();
 
                     lightManager->updateMaxLights(Settings::shaders().mMaxLights);
-
+/*
                     auto defines = mResourceSystem->getSceneManager()->getShaderManager().getGlobalDefines();
                     for (const auto& [name, key] : lightManager->getLightDefines())
                         defines[name] = key;
                     mResourceSystem->getSceneManager()->getShaderManager().setGlobalDefines(defines);
-
+*/
                     mStateUpdater->reset();
 
-                    mViewer->startThreading();
+//                    mViewer->startThreading();
                 }
             }
             else if (it->first == "Post Processing" && it->second == "enabled")
@@ -1824,4 +1838,9 @@ namespace MWRender
     {
         mNavMesh->setMode(value);
     }
+
+    void RenderingManager::setWriteNormals(bool enabled)
+    {
+        mStateUpdater->setWriteNormals(enabled);
+    }
 }
diff --git a/apps/openmw/mwrender/renderingmanager.hpp b/apps/openmw/mwrender/renderingmanager.hpp
index 4723c59b8a..622cd41056 100644
--- a/apps/openmw/mwrender/renderingmanager.hpp
+++ b/apps/openmw/mwrender/renderingmanager.hpp
@@ -276,6 +276,8 @@ namespace MWRender
 
         void setNavMeshMode(Settings::NavMeshRenderMode value);
 
+        void setWriteNormals(bool enabled);
+
     private:
         void updateTextureFiltering();
         void updateAmbient();
diff --git a/components/shader/shadervisitor.cpp b/components/shader/shadervisitor.cpp
index 0e6b5e79c6..2430a4f269 100644
--- a/components/shader/shadervisitor.cpp
+++ b/components/shader/shadervisitor.cpp
@@ -720,7 +720,10 @@ namespace Shader
             defineMap["endLight"] = "0";
 
         if (simpleLighting || dynamic_cast<osgParticle::ParticleSystem*>(&node))
+        {
             defineMap["forcePPL"] = "0";
+            writableStateSet->setDefine("FORCE_PPL", "0", osg::StateAttribute::ON | osg::StateAttribute::PROTECTED);
+        }
 
         bool particleOcclusion = false;
         node.getUserValue("particleOcclusion", particleOcclusion);
diff --git a/files/shaders/compatibility/groundcover.frag b/files/shaders/compatibility/groundcover.frag
index 96a79c8793..5b8f49b2bd 100644
--- a/files/shaders/compatibility/groundcover.frag
+++ b/files/shaders/compatibility/groundcover.frag
@@ -10,6 +10,8 @@
 
 #define GROUNDCOVER
 
+#pragma import_defines(WRITE_NORMALS, CLASSIC_FALLOFF, MAX_LIGHTS)
+
 #if @diffuseMap
 uniform sampler2D diffuseMap;
 varying vec2 diffuseMapUV;
@@ -85,7 +87,7 @@ void main()
     gl_FragData[0].xyz *= lighting;
     gl_FragData[0] = applyFogAtDist(gl_FragData[0], euclideanDepth, linearDepth, far);
 
-#if !@disableNormals
+#if defined(WRITE_NORMALS) && WRITE_NORMALS
     gl_FragData[1].xyz = viewNormal * 0.5 + 0.5;
 #endif
 
diff --git a/files/shaders/compatibility/groundcover.vert b/files/shaders/compatibility/groundcover.vert
index 8cf53a19e0..3849490dde 100644
--- a/files/shaders/compatibility/groundcover.vert
+++ b/files/shaders/compatibility/groundcover.vert
@@ -1,5 +1,7 @@
 #version 120
 
+#pragma import_defines(CLASSIC_FALLOFF, MAX_LIGHTS)
+
 #if @useUBO
     #extension GL_ARB_uniform_buffer_object : require
 #endif
diff --git a/files/shaders/compatibility/objects.frag b/files/shaders/compatibility/objects.frag
index 2e3ea795fb..3859c90e25 100644
--- a/files/shaders/compatibility/objects.frag
+++ b/files/shaders/compatibility/objects.frag
@@ -1,5 +1,5 @@
 #version 120
-#pragma import_defines(FORCE_OPAQUE, DISTORTION)
+#pragma import_defines(FORCE_OPAQUE, DISTORTION, WRITE_NORMALS, FORCE_PPL, CLASSIC_FALLOFF, MAX_LIGHTS)
 
 #if @useUBO
     #extension GL_ARB_uniform_buffer_object : require
@@ -68,7 +68,11 @@ uniform float far;
 uniform float alphaRef;
 uniform float distortionStrength;
 
+#if defined(FORCE_PPL)
+#define PER_PIXEL_LIGHTING (@normalMap || @specularMap || FORCE_PPL)
+#else
 #define PER_PIXEL_LIGHTING (@normalMap || @specularMap || @forcePPL)
+#endif
 
 #if !PER_PIXEL_LIGHTING
 centroid varying vec3 passLighting;
@@ -266,7 +270,7 @@ vec2 screenCoords = gl_FragCoord.xy / screenRes;
     gl_FragData[0].a = 1.0;
 #endif
 
-#if !defined(FORCE_OPAQUE) && !@disableNormals
+#if !defined(FORCE_OPAQUE) && defined(WRITE_NORMALS) && WRITE_NORMALS
     gl_FragData[1].xyz = viewNormal * 0.5 + 0.5;
 #endif
 
diff --git a/files/shaders/compatibility/objects.vert b/files/shaders/compatibility/objects.vert
index 081ff909cf..83b3d9ce48 100644
--- a/files/shaders/compatibility/objects.vert
+++ b/files/shaders/compatibility/objects.vert
@@ -1,5 +1,7 @@
 #version 120
 
+#pragma import_defines(FORCE_PPL, CLASSIC_FALLOFF, MAX_LIGHTS)
+
 #if @useUBO
     #extension GL_ARB_uniform_buffer_object : require
 #endif
@@ -49,7 +51,11 @@ varying vec2 specularMapUV;
 varying vec2 glossMapUV;
 #endif
 
+#if defined(FORCE_PPL)
+#define PER_PIXEL_LIGHTING (@normalMap || @specularMap || FORCE_PPL)
+#else
 #define PER_PIXEL_LIGHTING (@normalMap || @specularMap || @forcePPL)
+#endif
 
 #if !PER_PIXEL_LIGHTING
 centroid varying vec3 passLighting;
diff --git a/files/shaders/compatibility/terrain.frag b/files/shaders/compatibility/terrain.frag
index d206b9028f..5af540f269 100644
--- a/files/shaders/compatibility/terrain.frag
+++ b/files/shaders/compatibility/terrain.frag
@@ -8,6 +8,8 @@
     #extension GL_EXT_gpu_shader4: require
 #endif
 
+#pragma import_defines(WRITE_NORMALS, FORCE_PPL, CLASSIC_FALLOFF, MAX_LIGHTS)
+
 varying vec2 uv;
 
 uniform sampler2D diffuseMap;
@@ -23,7 +25,11 @@ uniform sampler2D blendMap;
 varying float euclideanDepth;
 varying float linearDepth;
 
+#if defined(FORCE_PPL)
+#define PER_PIXEL_LIGHTING (@normalMap || @specularMap || FORCE_PPL)
+#else
 #define PER_PIXEL_LIGHTING (@normalMap || @specularMap || @forcePPL)
+#endif
 
 #if !PER_PIXEL_LIGHTING
 centroid varying vec3 passLighting;
@@ -98,7 +104,7 @@ void main()
 
     gl_FragData[0] = applyFogAtDist(gl_FragData[0], euclideanDepth, linearDepth, far);
 
-#if !@disableNormals && @writeNormals
+#if defined(WRITE_NORMALS) && WRITE_NORMALS && @writeNormals
     gl_FragData[1].xyz = viewNormal * 0.5 + 0.5;
 #endif
 
diff --git a/files/shaders/compatibility/terrain.vert b/files/shaders/compatibility/terrain.vert
index 40d69ff35d..e0d8f2025e 100644
--- a/files/shaders/compatibility/terrain.vert
+++ b/files/shaders/compatibility/terrain.vert
@@ -1,5 +1,7 @@
 #version 120
 
+#pragma import_defines(FORCE_PPL, CLASSIC_FALLOFF, MAX_LIGHTS)
+
 #if @useUBO
     #extension GL_ARB_uniform_buffer_object : require
 #endif
@@ -13,7 +15,11 @@ varying vec2 uv;
 varying float euclideanDepth;
 varying float linearDepth;
 
+#if defined(FORCE_PPL)
+#define PER_PIXEL_LIGHTING (@normalMap || @specularMap || FORCE_PPL)
+#else
 #define PER_PIXEL_LIGHTING (@normalMap || @specularMap || @forcePPL)
+#endif
 
 #if !PER_PIXEL_LIGHTING
 centroid varying vec3 passLighting;
diff --git a/files/shaders/compatibility/water.frag b/files/shaders/compatibility/water.frag
index dbf36560b3..1ca387c9c3 100644
--- a/files/shaders/compatibility/water.frag
+++ b/files/shaders/compatibility/water.frag
@@ -8,6 +8,8 @@
     #extension GL_EXT_gpu_shader4: require
 #endif
 
+#pragma import_defines(WRITE_NORMALS, CLASSIC_FALLOFF, MAX_LIGHTS)
+
 #include "lib/core/fragment.h.glsl"
 
 // Inspired by Blender GLSL Water by martinsh ( https://devlog-martinsh.blogspot.de/2012/07/waterundewater-shader-wip.html )
@@ -245,7 +247,7 @@ void main(void)
 
     gl_FragData[0] = applyFogAtDist(gl_FragData[0], radialDepth, linearDepth, far);
 
-#if !@disableNormals
+#if defined(WRITE_NORMALS) && WRITE_NORMALS
     gl_FragData[1].rgb = normalize(gl_NormalMatrix * normal) * 0.5 + 0.5;
 #endif
 
diff --git a/files/shaders/lib/light/lighting.glsl b/files/shaders/lib/light/lighting.glsl
index 4a86233294..fd2696652e 100644
--- a/files/shaders/lib/light/lighting.glsl
+++ b/files/shaders/lib/light/lighting.glsl
@@ -66,7 +66,7 @@ void doLighting(vec3 viewPos, vec3 viewNormal, float shininess, out vec3 diffuse
         float lightDistance = length(lightPos);
 
         // cull non-FFP point lighting by radius, light is guaranteed to not fall outside this bound with our cutoff
-#if !@classicFalloff && !@lightingMethodFFP
+#if defined(CLASSIC_FALLOFF) && !CLASSIC_FALLOFF && !@lightingMethodFFP
         if (lightDistance > lcalcRadius(lightIndex) * 2.0)
             continue;
 #endif
diff --git a/files/shaders/lib/light/lighting_util.glsl b/files/shaders/lib/light/lighting_util.glsl
index b56afe285b..a749667905 100644
--- a/files/shaders/lib/light/lighting_util.glsl
+++ b/files/shaders/lib/light/lighting_util.glsl
@@ -35,7 +35,14 @@ struct LightData
     vec4 attenuation;
 };
 
+
+
+
+#if defined(MAX_LIGHTS)
+uniform int PointLightIndex[MAX_LIGHTS];
+#else
 uniform int PointLightIndex[@maxLights];
+#endif
 uniform int PointLightCount;
 
 // Defaults to shared layout. If we ever move to GLSL 140, std140 layout should be considered
@@ -54,7 +61,12 @@ uniform LightBufferBinding
 |  att_c  |  att_l   |  att_q   |  radius/spec_a  |
  --------------------------------------------------
 */
+
+#if defined(MAX_LIGHTS)
+uniform mat4 LightBuffer[MAX_LIGHTS];
+#else
 uniform mat4 LightBuffer[@maxLights];
+#endif
 uniform int PointLightCount;
 
 #endif
@@ -106,7 +118,7 @@ float lcalcRadius(int lightIndex)
 float lcalcIllumination(int lightIndex, float dist)
 {
     float illumination = 1.0 / (lcalcConstantAttenuation(lightIndex) + lcalcLinearAttenuation(lightIndex) * dist + lcalcQuadraticAttenuation(lightIndex) * dist * dist);
-#if !@classicFalloff && !@lightingMethodFFP
+#if defined(CLASSIC_FALLOFF) && !CLASSIC_FALLOFF && !@lightingMethodFFP
     // Fade illumination between the radius and the radius doubled to diminish pop-in
     illumination *= 1.0 - quickstep((dist / lcalcRadius(lightIndex)) - 1.0);
 #endif
